File: src/test/java/org/artembogomolova/demo/webapp/test/domain/entity/core/PhysicalAddressEntityTest.java
package org.artembogomolova.demo.webapp.test.domain.entity.core;

import lombok.EqualsAndHashCode;
import org.artembogomolova.demo.webapp.domain.core.PhysicalAddress;
import org.artembogomolova.demo.webapp.domain.core.PhysicalAddress_;
import org.artembogomolova.demo.webapp.test.domain.DomainTestUtil;
import org.artembogomolova.demo.webapp.test.domain.entity.AbstractAccessorEntityTest;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;

@DisplayName("Entity test: PhysicalAddress")
class PhysicalAddressEntityTest extends AbstractAccessorEntityTest<PhysicalAddress> {

  private static final String POSTAL_CODE_VALUE = "190000";
  private static final String COUNTRY_VALUE = "Russia";
  private static final String STATE_VALUE = "Saint Petersburg";
  private static final String CITY_VALUE = "Saint Petersburg";
  private static final String DISTRICT_VALUE = "District1";
  private static final String STREET_VALUE = "Street1";
  private static final String HOUSE_VALUE = "House1";
  private static final Integer ROOM_VALUE = 10;
  private static final String SPECIFIC_PART_VALUE = "SpecificPart1";

  PhysicalAddressEntityTest() {
    super(PhysicalAddress.class,
        PhysicalAddress::new,
        MockPhysicalAddress::new);
  }

  @Override
  protected PhysicalAddress buildStandardEntity() {
    PhysicalAddress result = new PhysicalAddress();
    result.setPostalCode(POSTAL_CODE_VALUE);
    result.setCountry(COUNTRY_VALUE);
    result.setState(STATE_VALUE);
    result.setCity(CITY_VALUE);
    result.setDistrict(DISTRICT_VALUE);
    result.setStreet(STREET_VALUE);
    result.setHouse(HOUSE_VALUE);
    result.setRoom(ROOM_VALUE);
    result.setSpecificPart(SPECIFIC_PART_VALUE);
    return result;
  }

  @Override
  protected void containFieldCorrectValuesTest(PhysicalAddress standardEntity) {
    Assertions.assertEquals(POSTAL_CODE_VALUE, standardEntity.getPostalCode());
    Assertions.assertEquals(COUNTRY_VALUE, standardEntity.getCountry());
    Assertions.assertEquals(STATE_VALUE, standardEntity.getState());
    Assertions.assertEquals(CITY_VALUE, standardEntity.getCity());
    Assertions.assertEquals(DISTRICT_VALUE, standardEntity.getDistrict());
    Assertions.assertEquals(STREET_VALUE, standardEntity.getStreet());
    Assertions.assertEquals(HOUSE_VALUE, standardEntity.getHouse());
    Assertions.assertEquals(ROOM_VALUE, standardEntity.getRoom());
    Assertions.assertEquals(SPECIFIC_PART_VALUE, standardEntity.getSpecificPart());
  }


  @Override
  protected PhysicalAddress buildAnotherEntityForTest() {
    return DomainTestUtil.buildTestAddress();
  }

  @Override
  protected boolean withoutBasicConstraint(PhysicalAddress standardEntity, String columnName) {
    if (isCountryPartColumnName(columnName)) {
      return withoutCountryPartEqualTest(standardEntity, columnName);
    }
    if (isCityLevelPartColumnName(columnName)) {
      return withoutCityPartEqualTest(standardEntity, columnName);
    }
    if (isHouseLevelPartColumnName(columnName)) {
      return withoutHousePartEqualTest(standardEntity, columnName);
    }
    return false;
  }


  private boolean withoutHousePartEqualTest(PhysicalAddress standardEntity, String columnName) {
    switch (columnName) {
      case PhysicalAddress_.HOUSE: {
        withoutColumnEqualTest(standardEntity, PhysicalAddress::getHouse, PhysicalAddress::setHouse);
        return true;
      }
      case PhysicalAddress_.ROOM: {
        withoutColumnEqualTest(standardEntity, PhysicalAddress::getRoom, PhysicalAddress::setRoom);
        return true;
      }
      case PhysicalAddress_.SPECIFIC_PART: {
        withoutColumnEqualTest(standardEntity, PhysicalAddress::getSpecificPart, PhysicalAddress::setSpecificPart);
        return true;
      }
      default: {
        return false;
      }
    }
  }

  private boolean isHouseLevelPartColumnName(String columnName) {
    return PhysicalAddress_.HOUSE.equals(columnName) || PhysicalAddress_.ROOM.equals(columnName) ||
        PhysicalAddress_.SPECIFIC_PART.equals(columnName);
  }

  private boolean withoutCityPartEqualTest(PhysicalAddress standardEntity, String columnName) {
    switch (columnName) {
      case PhysicalAddress_.CITY: {
        withoutColumnEqualTest(standardEntity, PhysicalAddress::getCity, PhysicalAddress::setCity);
        return true;
      }
      case PhysicalAddress_.DISTRICT: {
        withoutColumnEqualTest(standardEntity, PhysicalAddress::getDistrict, PhysicalAddress::setDistrict);
        return true;
      }
      case PhysicalAddress_.STREET: {
        withoutColumnEqualTest(standardEntity, PhysicalAddress::getStreet, PhysicalAddress::setStreet);
        return true;
      }
      default: {
        return false;
      }
    }
  }

  private boolean isCityLevelPartColumnName(String columnName) {
    return PhysicalAddress_.CITY.equals(columnName) || PhysicalAddress_.DISTRICT.equals(columnName) ||
        PhysicalAddress_.STREET.equals(columnName);
  }

  private boolean withoutCountryPartEqualTest(PhysicalAddress standardEntity, String columnName) {
    switch (columnName) {
      case PhysicalAddress_.POSTAL_CODE: {
        withoutColumnEqualTest(standardEntity, PhysicalAddress::getPostalCode, PhysicalAddress::setPostalCode);
        return true;
      }
      case PhysicalAddress_.COUNTRY: {
        withoutColumnEqualTest(standardEntity, PhysicalAddress::getCountry, PhysicalAddress::setCountry);
        return true;
      }
      case PhysicalAddress_.STATE: {
        withoutColumnEqualTest(standardEntity, PhysicalAddress::getState, PhysicalAddress::setState);
        return true;
      }
      default:
        return false;
    }
  }

  private boolean isCountryPartColumnName(String columnName) {
    return PhysicalAddress_.POSTAL_CODE.equals(columnName) || PhysicalAddress_.COUNTRY.equals(columnName) ||
        PhysicalAddress_.STATE.equals(columnName);
  }

  @EqualsAndHashCode(callSuper = false)
  private static class MockPhysicalAddress extends PhysicalAddress {

    public MockPhysicalAddress(PhysicalAddress physicalAddress) {
      super(physicalAddress);
    }
  }
}

File: app/src/test/kotlin/org/artembogomolova/demo/webapp/test/domain/core/PhysicalAddressEntityTest.kt
package org.artembogomolova.demo.webapp.test.domain.core

import org.artembogomolova.demo.webapp.main.domain.core.PhysicalAddress
import org.artembogomolova.demo.webapp.main.domain.core.PhysicalAddress_
import org.artembogomolova.demo.webapp.test.domain.AbstractAccessorEntityTest
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName

@DisplayName("Entity test: PhysicalAddress")
internal class PhysicalAddressEntityTest : AbstractAccessorEntityTest<PhysicalAddress>(
    PhysicalAddress::class.java,
    PhysicalAddress::from
) {
    override fun buildStandardEntity(): PhysicalAddress = PhysicalAddress(
        postalCode = POSTAL_CODE_VALUE,
        country = COUNTRY_VALUE,
        state = STATE_VALUE,
        city = CITY_VALUE,
        district = DISTRICT_VALUE,
        street = STREET_VALUE,
        house = HOUSE_VALUE,
        room = ROOM_VALUE,
        specificPart = SPECIFIC_PART_VALUE
    )


    override fun containFieldCorrectValuesTest(standardEntity: PhysicalAddress) {
        Assertions.assertEquals(POSTAL_CODE_VALUE, standardEntity.postalCode)
        Assertions.assertEquals(COUNTRY_VALUE, standardEntity.country)
        Assertions.assertEquals(STATE_VALUE, standardEntity.state)
        Assertions.assertEquals(CITY_VALUE, standardEntity.city)
        Assertions.assertEquals(DISTRICT_VALUE, standardEntity.district)
        Assertions.assertEquals(STREET_VALUE, standardEntity.street)
        Assertions.assertEquals(HOUSE_VALUE, standardEntity.house)
        Assertions.assertEquals(ROOM_VALUE, standardEntity.room)
        Assertions.assertEquals(SPECIFIC_PART_VALUE, standardEntity.specificPart)
    }

    override fun buildAnotherEntityForTest(): PhysicalAddress = PhysicalAddress(
        postalCode = "523152",
        country = "Country1",
        city = "City1",
        house = "42"
    )

    override fun withoutBasicConstraint(standardEntity: PhysicalAddress, columnName: String): Boolean {
        if (isCountryPartColumnName(columnName)) {
            return withoutCountryPartEqualTest(standardEntity, columnName)
        }
        if (isCityLevelPartColumnName(columnName)) {
            return withoutCityPartEqualTest(standardEntity, columnName)
        }
        return if (isHouseLevelPartColumnName(columnName)) {
            withoutHousePartEqualTest(standardEntity, columnName)
        } else false
    }

    private fun withoutHousePartEqualTest(standardEntity: PhysicalAddress, columnName: String): Boolean {
        return when (columnName) {
            PhysicalAddress_.HOUSE -> {
                withoutColumnEqualTest(standardEntity, PhysicalAddress::house)
                true
            }
            PhysicalAddress_.ROOM -> {
                withoutColumnEqualTest(standardEntity, PhysicalAddress::room)
                true
            }
            PhysicalAddress_.SPECIFIC_PART -> {
                withoutColumnEqualTest(standardEntity, PhysicalAddress::specificPart)
                true
            }
            else -> {
                false
            }
        }
    }

    private fun isHouseLevelPartColumnName(columnName: String): Boolean {
        return PhysicalAddress_.HOUSE == columnName || PhysicalAddress_.ROOM == columnName ||
                PhysicalAddress_.SPECIFIC_PART == columnName
    }

    private fun withoutCityPartEqualTest(standardEntity: PhysicalAddress, columnName: String): Boolean {
        return when (columnName) {
            PhysicalAddress_.CITY -> {
                withoutColumnEqualTest(standardEntity, PhysicalAddress::city)
                true
            }
            PhysicalAddress_.DISTRICT -> {
                withoutColumnEqualTest(standardEntity, PhysicalAddress::district)
                true
            }
            PhysicalAddress_.STREET -> {
                withoutColumnEqualTest(standardEntity, PhysicalAddress::street)
                true
            }
            else -> {
                false
            }
        }
    }

    private fun isCityLevelPartColumnName(columnName: String): Boolean {
        return PhysicalAddress_.CITY == columnName || PhysicalAddress_.DISTRICT == columnName ||
                PhysicalAddress_.STREET == columnName
    }

    private fun withoutCountryPartEqualTest(standardEntity: PhysicalAddress, columnName: String): Boolean {
        return when (columnName) {
            PhysicalAddress_.POSTAL_CODE -> {
                withoutColumnEqualTest(standardEntity, PhysicalAddress::postalCode)
                true
            }
            PhysicalAddress_.COUNTRY -> {
                withoutColumnEqualTest(standardEntity, PhysicalAddress::country)
                true
            }
            PhysicalAddress_.STATE -> {
                withoutColumnEqualTest(standardEntity, PhysicalAddress::state)
                true
            }
            else -> false
        }
    }

    private fun isCountryPartColumnName(columnName: String): Boolean {
        return PhysicalAddress_.POSTAL_CODE == columnName || PhysicalAddress_.COUNTRY == columnName ||
                PhysicalAddress_.STATE == columnName
    }

    companion object {
        private const val POSTAL_CODE_VALUE = "190000"
        private const val COUNTRY_VALUE = "Russia"
        private const val STATE_VALUE = "Saint Petersburg"
        private const val CITY_VALUE = "Saint Petersburg"
        private const val DISTRICT_VALUE = "District1"
        private const val STREET_VALUE = "Street1"
        private const val HOUSE_VALUE = "House1"
        private const val ROOM_VALUE = 10
        private const val SPECIFIC_PART_VALUE = "SpecificPart1"
    }
}